### DISCUSSION TITLE
Neotest Vitest

### DISCUSSION THREAD

**Discussion by marilari88** [nvim-neotest/neotest#96] [Show and tell]
I recently worked on a JS project using Vitest as test framework and I felt uncomfortable that I couldn't use Neotest 😂
As a result, I spent my last weekend setting up the vitest adapter. This is based on [neotest-jest](https://github.com/haydenmeade/neotest-jest) and I worked on changing a few things regarding test command and parsing

Here is the repo:
https://github.com/marilari88/neotest-vitest

Currently, It seems to work fine for my dev experience (it solves the three problems: parse tests, construct test commands and collect results) but there are still a few little issues I will work on. Contributions are very appreciate

Special thanks to @rcarriga, and all contributors for working on a such an amazing test framework

**Comment by rcarriga** [comment]
Fantastic to see thanks for creating an adapter!

Had a skim through and saw
> Weird behaviors when either neotest-jest is installed at same time (they share same test parsing logic)

Could you elaborate on these issues? Is it something going wrong in neotest core or just the adapter?

**Reply by marilari88** [reply]
> Could you elaborate on these issues? Is it something going wrong in neotest core or just the adapter?

I try to be more specific:
- Working on a JS project and having neotest-jest and neotest-vitest enabled in my config,  Neotest shows two trees in the summary panel (even if jest or vitest are not installed in the project).
- Pressing `e` on the second listed tree namespaces opens the corresponding details of the first one
- NeotestNearest, NeotestFile, Neotest take into account only the first listed adapter (no way to run test with the second one with through commands)

Maybe we should find a way to look into the package.json in order to enable the right adapter.

**Reply by rcarriga** [reply]
Ah OK so the first issue can be resolved by jest and vitest checking if they're installed and not returning a root from the `root` adapter function. That way only one should show up for most people (assuming they don't have both installed).

The second issue is a neotest core issue, thanks for raising!

Those commands are used defined I'm assuming, since neotest doesn't define them. The run consumer could provide some way to query which adapter to use in the case of two adapters matching a position. One thought is that neotest.run could accept the `adapter` argument as a function to call to select one if multiple are found.

**Comment by tyler-dot-earth** [comment]
I was setting up neotest for the first time today and stumbled on this thread - managed to get [marilari88/neotest-vitest](https://github.com/marilari88/neotest-vitest) working just fine :+1: Thanks for making this, I'll contribute with feedback/issues if nothing else.
